crates/uv-distribution/src/distribution_database.rs

```diff
@@ -342,8 +342,21 @@ impl<'a, Context: BuildContext + Send + Sync> DistributionDatabase<'a, Context>
     pub async fn get_wheel_metadata(
         &self,
         dist: &BuiltDist,
-        hashes: &[HashDigest],
+        hashes: HashPolicy<'_>,
     ) -> Result<ArchiveMetadata, Error> {
+        // If hash generation is enabled, and the distribution isn't hosted on an index, get the
+        // entire wheel to ensure that the hashes are included in the response. If the distribution
+        // is hosted on an index, the hashes will be included in the simple metadata response.
+        // For hash _validation_, callers are expected to enforce the policy when retrieving the
+        // wheel.
+        // TODO(charlie): Request the hashes via a separate method, to reduce the coupling in this API.
+        if hashes.is_generate() && matches!(dist, BuiltDist::DirectUrl(_) | BuiltDist::Path(_)) {
+            let wheel = self.get_wheel(dist, hashes).await?;
+            let metadata = wheel.metadata()?;
+            let hashes = wheel.hashes;
+            return Ok(ArchiveMetadata { metadata, hashes });
+        }
```

Comment:
Worst part of this PR but not 100% sure how to solve. It's fine in practice, just somewhat brittle.